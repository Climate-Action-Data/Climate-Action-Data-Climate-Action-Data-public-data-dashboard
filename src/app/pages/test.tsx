'use client'
import { type NextPage } from "next";

const Test: NextPage = () => {
    return (
        <div>
            <h1>Hello</h1>
            <p>data</p>
        </div>
    )
}

export default Test