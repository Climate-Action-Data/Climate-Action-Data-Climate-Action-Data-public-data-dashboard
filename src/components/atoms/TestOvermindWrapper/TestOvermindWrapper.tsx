import { createOvermind } from 'overmind'
import { Provider } from 'overmind-react'
import { config } from '@/overmind'
import { PropsWithChildren } from 'react'

const overmind = createOvermind(config)

export const TestOvermindWrapper = (props: PropsWithChildren) => {
  return <Provider value={overmind}>{props.children}</Provider>
}
