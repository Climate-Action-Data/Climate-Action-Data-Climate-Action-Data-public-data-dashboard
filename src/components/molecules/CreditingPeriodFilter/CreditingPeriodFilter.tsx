import { Button, Flex, HStack, Popover, PopoverBody, PopoverContent, PopoverTrigger, Spacer, VStack } from '@chakra-ui/react'
import { DropDownIcon } from '@/components/atoms/DropDownIcon/DropDownIcon'
import { Dispatch, FC, SetStateAction, useEffect, useState } from 'react'
import CalendarWrapper from '@/components/molecules/CalendarWrapper/CalendarWrapper'
import { MinusIcon } from '@/components/atoms/MinusIcon/MinusIcon'
import DateInput from '@/components/molecules/DateInput/DateInput'
import { useTranslation } from 'react-i18next'
import { format } from 'date-fns'
import { DateFormats } from '@/@types/DateFormats'
import { DatesFilter } from '@/@types/ProjectSearchFilterValues'

interface CreditingPeriodFilterProp {
  label: string
  earliestDate?: Date
  latestDate?: Date
  dateFilter?: DatesFilter
  onDateChange: (payload: DatesFilter) => void
}

const CreditingPeriodFilter: FC<CreditingPeriodFilterProp> = (prop) => {
  const { label, earliestDate, latestDate, onDateChange, dateFilter } = prop

  const [minimumDate, setMinimumDate] = useState<Date | undefined>()
  const [maximumDate, setMaximumDate] = useState<Date | undefined>()
  const [showMinimumDateSelector, setShowMinimumDateSelector] = useState<boolean>(false)
  const [showMaximumDateSelector, setShowMaximumDateSelector] = useState<boolean>(false)

  const { t } = useTranslation(`search`)

  useEffect(() => {
    setMinimumDate(dateFilter?.minDate)
    setMaximumDate(dateFilter?.maxDate)
  }, [dateFilter])

  const getOnApplySelectedDate = (action: Dispatch<SetStateAction<Date | undefined>>, selectedDate: Date | undefined) => {
    action(selectedDate)
    setShowMinimumDateSelector(false)
    setShowMaximumDateSelector(false)
  }

  const renderDateSelector = (stateValue: Date | undefined, action: Dispatch<SetStateAction<Date | undefined>>) => {
    return (
      <CalendarWrapper
        preSelectedDate={stateValue}
        maxDate={latestDate ?? maximumDate}
        minDate={earliestDate ?? minimumDate}
        onApplySelectedDate={(selectedDate) => getOnApplySelectedDate(action, selectedDate)}
      />
    )
  }

  const handleOnOpenDatePickerMinimumDate = () => {
    setShowMinimumDateSelector(true)
  }

  const handleOnOpenDatePickerMaximumDate = () => {
    setShowMaximumDateSelector(true)
  }

  const handleOnClearClick = () => {
    setMinimumDate(undefined)
    setMaximumDate(undefined)
    onDateChange({})
  }

  const handleOnApplyClick = () => {
    if (minimumDate && maximumDate) {
      if (minimumDate <= maximumDate) {
        onDateChange({ maxDate: maximumDate, minDate: minimumDate })
      }
    }
    if (minimumDate) {
      onDateChange({ minDate: minimumDate })
    }
    if (maximumDate) {
      onDateChange({ maxDate: maximumDate })
    }
  }

  const handleOnClose = () => {
    setShowMinimumDateSelector(false)
    setShowMaximumDateSelector(false)
  }

  const handleMinDateOnChange = (date: Date | undefined) => {
    setMinimumDate(date)
  }

  const handleMaxDateOnChange = (date: Date | undefined) => {
    setMaximumDate(date)
  }

  const renderContent = () => {
    if (showMaximumDateSelector || showMinimumDateSelector) {
      return null
    }
    return (
      <VStack>
        <HStack width={`100%`} justify={`center`} alignItems={`end`}>
          <DateInput
            label={`Minimum Date`}
            value={minimumDate}
            onOpenDatePicker={handleOnOpenDatePickerMinimumDate}
            maxDate={latestDate ?? maximumDate}
            minDate={earliestDate}
            onChange={handleMinDateOnChange}
          />
          <MinusIcon />
          <DateInput
            label={`Maximum Date`}
            value={maximumDate}
            onOpenDatePicker={handleOnOpenDatePickerMaximumDate}
            maxDate={latestDate}
            minDate={earliestDate ?? minimumDate}
            onChange={handleMaxDateOnChange}
          />
        </HStack>
        <HStack justify={`space-between`} width={`100%`} margin={`8px`}>
          <Button variant={`textLink`} onClick={handleOnClearClick} data-testid={`crediting-period-filter-clear`}>
            {t(`clear`)}
          </Button>
          <Button variant={`textLink`} onClick={handleOnApplyClick} data-testid={`crediting-period-filter-apply`}>
            {t(`apply`)}
          </Button>
        </HStack>
      </VStack>
    )
  }

  const renderLabel = () => {
    if (minimumDate && maximumDate) {
      return `${format(minimumDate, DateFormats.YYYY_MM_DD)} - ${format(maximumDate, DateFormats.YYYY_MM_DD)}`
    }
    if (minimumDate) {
      return `${format(minimumDate, DateFormats.YYYY_MM_DD)} and later`
    }
    if (maximumDate) {
      return t(`upTo`, { date: format(maximumDate, DateFormats.YYYY_MM_DD) })
    }
    return label
  }

  return (
    <Popover gutter={0} isLazy placement={`bottom`} matchWidth onClose={handleOnClose}>
      <PopoverTrigger>
        <Button variant={`dropdownUnselected`}>
          <Flex fontWeight={`normal`} fontSize={`16px`} alignItems={`center`} grow={1}>
            {renderLabel()}
            <Spacer />
            <DropDownIcon />
          </Flex>
        </Button>
      </PopoverTrigger>
      <PopoverContent zIndex={401} border={0} borderRadius={`8px`} width={[`100%`, null, null, `min-content`]} height={`min-content`} backgroundColor={`white`} boxShadow={`xl`}>
        <PopoverBody width={`100%`}>
          {renderContent()}
          {showMinimumDateSelector && renderDateSelector(minimumDate, setMinimumDate)}
          {showMaximumDateSelector && renderDateSelector(maximumDate, setMaximumDate)}
        </PopoverBody>
      </PopoverContent>
    </Popover>
  )
}

export default CreditingPeriodFilter
