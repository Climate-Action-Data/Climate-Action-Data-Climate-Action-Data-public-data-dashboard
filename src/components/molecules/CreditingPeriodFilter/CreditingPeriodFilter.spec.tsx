import { render, screen } from '@testing-library/react'
import CreditingPeriodFilter from '@/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter'
import { TestOvermindWrapper } from '@/test/TestOvermindWrapper'
import userEvent from '@testing-library/user-event'
import { MockData } from '@/test/TestOvermindMockData'
import { format } from 'date-fns'

describe(`CreditingPeriodFilter`, () => {
  const tYear = 2023
  const tMonth = 7
  const tDay = 1
  const tDate = new Date(tYear, tMonth, tDay)
  const tLabel = `label`
  const user = userEvent.setup({ delay: null })

  beforeAll(() => {
    jest.useFakeTimers()
    jest.setSystemTime(tDate)
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  test(`renders a CreditingPeriodFilter and open it`, async () => {
    const { container } = render(
      <TestOvermindWrapper>
        <CreditingPeriodFilter label={tLabel} />
      </TestOvermindWrapper>,
    )

    await user.click(screen.getByText(tLabel))
    expect(container).toMatchSnapshot()
  })

  test(`renders a CreditingPeriodFilter with a pre-selected max date`, async () => {
    const mockData = MockData.SEARCH_FILTER_VALUES
    mockData.selectedSearchFilterValues.searchFilterValues.filterDates = { maxDate: tDate }

    const tLabel2 = `Up to ${format(tDate, `dd/MM/yyyy`)}`

    const { container } = render(
      <TestOvermindWrapper searchFilters={mockData}>
        <CreditingPeriodFilter label={tLabel} />
      </TestOvermindWrapper>,
    )

    await user.click(screen.getByText(tLabel2))
    await user.click(screen.getByTestId(`crediting-period-filter-apply`))

    expect(container).toMatchSnapshot()
  })

  test(`renders a CreditingPeriodFilter with a pre-selected min date and click on clear`, async () => {
    const mockData = MockData.SEARCH_FILTER_VALUES
    mockData.selectedSearchFilterValues.searchFilterValues.filterDates = { minDate: tDate }

    const tLabel2 = `${format(tDate, `dd/MM/yyyy`)} and later`

    const { container } = render(
      <TestOvermindWrapper searchFilters={mockData}>
        <CreditingPeriodFilter label={tLabel} />
      </TestOvermindWrapper>,
    )

    await user.click(screen.getByText(tLabel2))
    await user.click(screen.getByTestId(`crediting-period-filter-clear`))

    expect(container).toMatchSnapshot()
  })

  test(`renders a CreditingPeriodFilter with a pre-selected min date and max date and click on apply`, async () => {
    const mockData = MockData.SEARCH_FILTER_VALUES
    mockData.selectedSearchFilterValues.searchFilterValues.filterDates = { minDate: tDate, maxDate: tDate }

    const tLabel2 = `${format(tDate, `dd/MM/yyyy`)} - ${format(tDate, `dd/MM/yyyy`)}`

    const { container } = render(
      <TestOvermindWrapper searchFilters={mockData}>
        <CreditingPeriodFilter label={tLabel} />
      </TestOvermindWrapper>,
    )

    await user.click(screen.getByText(tLabel2))
    await user.click(screen.getByTestId(`crediting-period-filter-apply`))

    expect(container).toMatchSnapshot()
  })

  test(`renders a CreditingPeriodFilter and open the date selectors and click on clear on the calendar`, async () => {
    const { container } = render(
      <TestOvermindWrapper>
        <CreditingPeriodFilter label={tLabel} />
      </TestOvermindWrapper>,
    )

    const tRenderedLabel = screen.getByText(tLabel)

    await user.click(tRenderedLabel)

    const tDateSelectors = screen.getAllByTestId(`datepicker-trigger`)

    // eslint-disable-next-line no-magic-numbers
    expect(tDateSelectors.length).toBe(2)

    await user.click(tDateSelectors[0])
    await user.type(tDateSelectors[0], `{esc}`)
    await user.click(tRenderedLabel)

    const tDateSelectors2 = screen.getAllByTestId(`datepicker-trigger`)

    await user.click(tDateSelectors2[1])
    await user.click(screen.getByTestId(`calender-wrapper-clear`))

    expect(container).toMatchSnapshot()
  })
})
