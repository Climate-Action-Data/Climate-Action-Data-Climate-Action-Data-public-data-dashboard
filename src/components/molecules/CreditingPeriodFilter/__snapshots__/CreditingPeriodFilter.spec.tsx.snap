// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditingPeriodFilter renders correctly 1`] = `
<div>
  <button
    aria-controls="popover-content-:r1:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="chakra-button css-4xx2wk"
    id="popover-trigger-:r1:"
    type="button"
  >
    <div
      class="css-jmqhys"
    >
      label
      <div
        class="css-17xejub"
      />
      <svg
        class="chakra-icon css-c0air1"
        focusable="false"
        viewBox="0 0 8 4"
      >
        <path
          d="M3.99911 3.9034L0.191406 0.0957031H7.80681L3.99911 3.9034Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </button>
  <div
    class="chakra-popover__popper css-0"
    style="visibility: hidden; position: absolute; inset: 0 auto auto 0;"
  >
    <section
      class="chakra-popover__content css-vbzucn"
      id="popover-content-:r1:"
      role="dialog"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
      tabindex="-1"
    />
  </div>
</div>
`;
